### ISSUE TITLE
New session: data augmentation on CIFAR-10 (`sessions/augmentation.ipynb`)

### ISSUE THREAD

**Issue by yoavram** [yoavram/DataSciPy#9]
**Branch:** `DL2026` · **New file:** `sessions/augmentation.ipynb` (+ `solutions/augmentation.ipynb`)

## Motivation

The branch has no augmentation anywhere: zero uses of `keras.layers.Random*` across `sessions/`, `exercises/` and `solutions/`. It is mentioned twice as an explanation for results we do not have — `transfer.ipynb` cell 43 ("A horizontal flip augmentation is worth another 0.4%") and `audio.ipynb` cell 63 ("heavier augmentation") — so students meet the word without ever meeting the mechanism.

It cannot be retrofitted into an existing notebook. MNIST is the wrong dataset (h-flip destroys 2/5/7, rotation merges 6/9, and at 60k images with 99% validation accuracy there is nothing to gain); the transfer and audio sessions are dominated by pretrained features, and `transfer.ipynb`'s probe caches embeddings once, which makes augmentation structurally impossible there.

CIFAR-10 is the smallest standard dataset where the label genuinely survives the transformations (no digits or text, so horizontal flip is exactly label-preserving) **and** a small CNN overfits visibly. This is also the only session on the branch where a from-scratch image model is shown overfitting and then fixed.

## Placement

New session in `index.ipynb`, Day 3, immediately after `[Convolutional networks with Keras](sessions/K_CNN.ipynb)`:

```
- [Data augmentation](sessions/augmentation.ipynb) | [solution](solutions/augmentation.ipynb)
```

Budget ~20-25 minutes in class, with all training cells commented out and checkpoints loaded.

## Data

`keras.datasets.cifar10.load_data()` — no TensorFlow, same route as the MNIST already used in `K_CNN.ipynb`. ~170 MB into the Keras cache.

- Add a `_cifar10()` entry to `KERAS_ITEMS` in `download_data.py`, described as `"CIFAR-10 (Keras cache) - sessions/augmentation.ipynb (Day 3)"`.
- Normalize to `[0, 1]` (`astype('float32') / 255`), one-hot the labels, shapes `(50000, 32, 32, 3)` and `(10000, 32, 32, 3)`.
- Use the official 10k test split as `X_validation` / `Y_validation`, and **validation** terminology throughout, per house style. Add one markdown sentence noting that nothing is selected on it — no early stopping, no epoch choice, one seed — so it is not being used as a selection set.
- A figure of 10 example images with class names, one per class, early in the notebook.

## Model

```python
def build_model(augment=None, seed=SEED):
    keras.utils.set_random_seed(seed)
    inputs = keras.Input(shape=(32, 32, 3))
    x = augment(inputs) if augment is not None else inputs
    for filters in (32, 64, 128):
        x = keras.layers.Conv2D(filters, 3, activation='relu', padding='same')(x)
        x = keras.layers.MaxPool2D()(x)
    x = keras.layers.Flatten()(x)
    x = keras.layers.Dense(256, activation='relu')(x)
    x = keras.layers.Dropout(rate=0.5)(x)
    outputs = keras.layers.Dense(ncats, activation='softmax')(x)
    return keras.Model(inputs, outputs, name='cifar_cnn')
```

620,362 parameters. Three conv blocks rather than two wider ones: with two blocks and 64/128 filters, ~93% of the parameters sit in the `Flatten→Dense` matrix and the receptive field never covers the image. Keep the `Dropout` — it is the regularizer students already met in `K_CNN.ipynb`, and keeping it makes the augmentation gain an honest increment on top of it rather than a comparison against an unregularized straw man.

The `augment` argument is deliberately a parameter, so the in-notebook exercise is a one-cell change.

**No BatchNorm, no learning-rate schedule, no callbacks, no early stopping.** `optimizer='adam'`, `loss='categorical_crossentropy'`, `metrics=['accuracy']`, `batch_size=64`.

## The experiment

Two arms, one seed, 50 epochs each, fixed budget:

```python
augment_simple = keras.Sequential([
    keras.layers.RandomFlip('horizontal'),
    keras.layers.RandomTranslation(0.125, 0.125, fill_mode='constant', fill_value=0.0),
], name='augment')
```

Flip plus ±4 px translation is the standard CIFAR recipe (`0.125 * 32 = 4`), deliberately minimal so the exercise has somewhere to go.

**Three API traps that must be handled explicitly in code and named in the markdown** — each one fails silently:

1. `RandomFlip` defaults to `mode='horizontal_and_vertical'`. Vertical flip is not a symmetry of CIFAR-10. Pass `'horizontal'`.
2. `RandomTranslation` defaults to `fill_mode='reflect'`. The standard recipe pads with zeros; pass `fill_mode='constant', fill_value=0.0`.
3. `RandomErasing`, `RandomContrast` and `RandomBrightness` default to `value_range=(0, 255)` while this notebook normalizes to `[0, 1]`. They must be given `value_range=(0, 1)`. This matters for the exercise, so state it before the exercise, not after.

The augmentation block goes **inside** the model, not applied to the arrays. Two consequences worth a markdown sentence: Keras runs these layers only when `training=True`, so validation is never augmented and inference is deterministic; and the layers are saved inside the `.keras` file, where they are inert.

### Live cell (runs on CPU in seconds)

Before any training: take one training image, call `augment_simple(x, training=True)` eight times, and plot the original plus the eight variants in a row. This is the only cell that must execute live in class, and it is what makes the mechanism concrete.

### Training and checkpoints (house pattern)

Follow the `K_CNN.ipynb` / `transfer.ipynb` idiom exactly: the `fit` cell, then the save cell, then the load cell. After the GPU run, comment out the `fit` cells so the committed notebook executes through the load path.

```
../data/cifar10_noaug.keras          ../data/cifar10_noaug_history.p
../data/cifar10_augmented.keras      ../data/cifar10_augmented_history.p
../data/cifar10_size_sweep.p
```

All of these are gitignored by extension (`*.keras`, `*.p`); do not commit them. Notebook outputs *are* committed, per `CLAUDE.md`.

`history = model.fit(...).history` unwrapped at the call site, as elsewhere.

### What to plot and report

1. Both arms' accuracy and loss curves, using the local `plot_history` helper style from the other Keras sessions — but **the headline figure is train-minus-validation accuracy versus epoch for both arms**, on one axis. The claim is about regularization, so the gap is the observable; the accuracy delta is only the headline number.
2. A small table: final-epoch training accuracy, final-epoch validation accuracy, and the gap, for both arms.
3. Report **final-epoch** numbers, not best-epoch. With no early stopping this is the honest comparison, and if the unaugmented arm's validation accuracy peaks early and then drifts down while its training accuracy goes to ~99%, that is the lesson — do not "fix" it, do not add a callback, do not report its peak.

Rough expectation from the literature for a net this size: unaugmented ~70% validation with training accuracy near 99%; augmented low-to-mid 70s or better, still improving at epoch 50, with a much smaller gap. **These are expectations, not targets. Report what the run produces.** If augmentation helps by less than expected, say so and note that a 620k-parameter net is partly capacity-limited — a measured disappointment is in keeping with the rest of this branch and is better teaching than a tuned result.

One markdown paragraph should state the confound plainly: augmentation slows fitting, so at any fixed epoch budget the comparison understates it, and the augmented curve has not converged at 50 epochs. Single seed, so differences below ~1% (binomial SE on 10k validation images is ~0.45%) should not be interpreted.

## In-notebook exercise

**Task:** extend `augment_simple` with `RandomRotation`, `RandomContrast`, `RandomErasing` (and optionally `RandomZoom`), retrain, and report whether the extra augmentations help.

Give students:
- the `value_range=(0, 1)` requirement,
- a warning that CIFAR objects are upright, so rotation should stay small (`factor≈0.05`, i.e. ±18°) and that large rotations will hurt,
- the observation that `RandomErasing` is dropout in input space with spatial structure — connect it to the `Dropout` layer already in the model,
- a reduced budget (e.g. 20 epochs) so the exercise is feasible.

`solutions/augmentation.ipynb` must be a working twin per `CLAUDE.md`, with its own checkpoint and history (`../data/cifar10_augmented_heavy.keras`, `..._history.p`) produced in the same GPU run. My expectation is that the heavy recipe adds little or nothing at this model size; if so, the solution notebook should say that and explain why, rather than being rewritten until it wins.

## Final section: accuracy versus training-set size

The point of the session that a single A/B cannot show — augmentation's value shrinks as data grows.

- Training-set sizes `N ∈ {2500, 5000, 10000, 25000, 50000}`, both arms, same subset indices for both arms, drawn with a fixed `np.random.default_rng(SEED)` permutation. Check the class balance of each subset and mention it.
- **Equal gradient-step budget, not equal epochs**: `steps = 50 * 50000 / 64 ≈ 39,062`, so `epochs = ceil(steps * 64 / N)` for each size. This makes the `N = 50,000` point identical to the headline run above, so it can reuse those two checkpoints — 8 new runs, not 10. Fixed epochs instead would confound data size with optimization budget.
- Plot: validation accuracy versus `N` on a log x-axis, two curves, plus the gap between them as the quantity of interest.
- Pickle the results to `../data/cifar10_size_sweep.p` and follow the same commented-out-training / load pattern as `autoencoders.ipynb` cell 46-51.
- One markdown sentence on framing: augmentation is not "more data". The sample size is unchanged; what changes is the hypothesis class, constrained toward functions invariant to a chosen transformation. That is why the curves converge as `N` grows.

## Cross-references

- Add a line to `transfer.ipynb` cell 43 pointing the orphaned "+0.4% from horizontal flip" at this session.
- Mention in the discussion that `audio.ipynb`'s 17 overlapping segments per clip are already augmentation, and that its segment-score averaging is test-time augmentation.
- Optional, cheap, and a good closer: TTA on the augmented model — average the softmax over the image and its horizontal flip, report the change.

## Execution and acceptance

Two-stage GPU run:

1. Execute with the `fit` cells live to produce the checkpoints, histories and sweep pickle.
2. Comment out every `fit` cell, then execute end to end again so the committed outputs come from the load path and the notebook runs fast for a student who has the artifacts.

Acceptance criteria:

- `sessions/augmentation.ipynb` and `solutions/augmentation.ipynb` exist, both linked from `index.ipynb`, all links resolve on `DL2026`.
- House style: `![Py4Eng](../logo.png)` opener, title, `## Yoav Ram`, "In this session we will understand:" list, first code cell with `%matplotlib inline` + imports + the Keras/backend/JAX print line, fixed `SEED`, `seaborn.set(style='white', context='talk')`, markdown narrative between every code cell, closing `# References` and `# Colophon` copied verbatim from a sibling.
- References to include: Krizhevsky 2009 (CIFAR-10), Shorten & Khoshgoftaar 2019 (survey), DeVries & Taylor 2017 (Cutout), Zhong et al. 2020 (Random Erasing), and the Keras image-augmentation layer docs.
- `nbconvert --execute` runs clean with the training cells commented out; no unintended errors.
- No new dependencies. All layers used are present in `keras==3.15.1` (verified: `RandomFlip`, `RandomTranslation`, `RandomRotation`, `RandomZoom`, `RandomContrast`, `RandomErasing`, plus `RandAugment`/`CutMix`/`MixUp`), and in-model augmentation trains, saves and reloads correctly on the JAX backend.
- `download_data.py --list` shows the CIFAR-10 entry.
- No checkpoints, pickles or arrays committed.

Measured cost on one vCPU for this architecture at batch 64: ~2.2 min/epoch without augmentation, ~3.0 min/epoch with flip + translation (1.4× overhead). So the full protocol is GPU-only; on a student laptop the notebook is expected to run only through the load path, and the augmentation-visualization cell is the only one that must execute live.

**Comment by yoavram** [comment]
Implemented in branch `augmentation`.

**Comment by yoavram** [comment]
Done on `DL2026`, merged in 00f7c15.

- 0a2f238 — add a data augmentation session on CIFAR-10
- e6bc1cc — add the accuracy-versus-training-set-size sweep
- 1eed651 — address the review: cost figures, and an ablation that overturns the exercise's conclusion

`sessions/augmentation.ipynb` and `solutions/augmentation.ipynb` are in, linked from `index.ipynb` on Day 3 immediately after Convolutional networks with Keras. The CIFAR-10 entry is in `download_data.py`, and `transfer.ipynb`'s orphaned "+0.4% from horizontal flip" remark now points here.

Built as specified: the three-block CNN with `augment` as a constructor parameter, augmentation inside the model, final-epoch numbers, train-minus-validation gap as the headline figure, and the size sweep at equal gradient-step budget so `N = 50,000` reuses the headline checkpoints. All three silent-failure traps (`RandomFlip` defaulting to `horizontal_and_vertical`, `RandomTranslation` defaulting to `fill_mode='reflect'`, `value_range=(0, 255)` against a `[0, 1]` notebook) are handled in code and named in the markdown before the exercise.

The expectation in the issue that the heavy recipe would add little at this model size held, and the solution notebook says so rather than being tuned until it won.
